test/test_netcdf_output_writer.jl

```diff
@@ -46,133 +55,1427 @@ function test_DateTime_netcdf_output(arch)
+            filename = filepath_sliced,
+            indices = (:, j_slice, k_slice),
+            schedule = IterationInterval(1),
+            array_type = Array{Float64},
+            with_halos = false,
+            include_grid_metrics = true,
+            verbose = true
+        )
+
+    run!(simulation)
+
+    # Test NetCDF output with halos
+    ds_h = NCDataset(filepath_with_halos)
+
+    dims = ("y_f", "y_c", "z_f", "z_c")
+    not_dims = ("x_f", "x_c")
+
+    metrics = ("dy_f", "dy_c", "dz_f", "dz_c")
+    not_metrics = ("dx_f", "dx_c")
+
+    vars = ("u", "v", "w", "T", "S")
+
+    for var in (dims..., metrics..., vars...)
+        @test haskey(ds_h, var)
+        @test haskey(ds_h[var].attrib, "long_name")
+        @test haskey(ds_h[var].attrib, "units")
+        @test eltype(ds_h[var]) == Float64
+    end
+
+    for var in (not_dims..., not_metrics...)
+        @test !haskey(ds_h, var)
+    end
+
+    @test dimsize(ds_h[:y_f]) == (y_f=Ny + 2Hy,)
+    @test dimsize(ds_h[:y_c]) == (y_c=Ny + 2Hy,)
+    @test dimsize(ds_h[:z_f]) == (z_f=Nz + 2Hz + 1,)
+    @test dimsize(ds_h[:z_c]) == (z_c=Nz + 2Hz,)
+
+    @test dimsize(ds_h[:dy_f]) == (y_f=Ny + 2Hy,)
+    @test dimsize(ds_h[:dy_c]) == (y_c=Ny + 2Hy,)
+    @test dimsize(ds_h[:dz_f]) == (z_f=Nz + 2Hz + 1,)
+    @test dimsize(ds_h[:dz_c]) == (z_c=Nz + 2Hz,)
+
+    @test dimsize(ds_h[:u]) == (y_c=Ny + 2Hy, z_c=Nz + 2Hz,     time=Nt + 1)
+    @test dimsize(ds_h[:v]) == (y_f=Ny + 2Hy, z_c=Nz + 2Hz,     time=Nt + 1)
+    @test dimsize(ds_h[:w]) == (y_c=Ny + 2Hy, z_f=Nz + 2Hz + 1, time=Nt + 1)
+    @test dimsize(ds_h[:T]) == (y_c=Ny + 2Hy, z_c=Nz + 2Hz,     time=Nt + 1)
+    @test dimsize(ds_h[:S]) == (y_c=Ny + 2Hy, z_c=Nz + 2Hz,     time=Nt + 1)
+
+    close(ds_h)
+    rm(filepath_with_halos)
+
+    # Test NetCDF sliced output
+    ds_s = NCDataset(filepath_sliced)
+
+    for var in (dims..., metrics..., vars...)
+        @test haskey(ds_s, var)
+        @test haskey(ds_s[var].attrib, "long_name")
+        @test haskey(ds_s[var].attrib, "units")
+        @test eltype(ds_s[var]) == Float64
+    end
+
+    for var in (not_dims..., not_metrics...)
+        @test !haskey(ds_s, var)
+    end
+
+    @test dimsize(ds_s[:y_f]) == (y_f=ny,)
+    @test dimsize(ds_s[:y_c]) == (y_c=ny,)
+    @test dimsize(ds_s[:z_f]) == (z_f=nz,)
+    @test dimsize(ds_s[:z_c]) == (z_c=nz,)
+
+    @test dimsize(ds_s[:dy_f]) == (y_f=ny,)
+    @test dimsize(ds_s[:dy_c]) == (y_c=ny,)
+    @test dimsize(ds_s[:dz_f]) == (z_f=nz,)
+    @test dimsize(ds_s[:dz_c]) == (z_c=nz,)
+
+    @test dimsize(ds_s[:u]) == (y_c=ny, z_c=nz, time=Nt + 1)
+    @test dimsize(ds_s[:v]) == (y_f=ny, z_c=nz, time=Nt + 1)
+    @test dimsize(ds_s[:w]) == (y_c=ny, z_f=nz, time=Nt + 1)
+    @test dimsize(ds_s[:T]) == (y_c=ny, z_c=nz, time=Nt + 1)
+    @test dimsize(ds_s[:S]) == (y_c=ny, z_c=nz, time=Nt + 1)
+
+    close(ds_s)
+    rm(filepath_sliced)
+
+    return nothing
+end
+
+function test_netcdf_rectilinear_column(arch)
```

Comment:
I see that the tests are pretty thorough! Well done!

I guess my only comment is that it seems we're re-writing a bunch of code. For example, `test_netcdf_rectilinear_flat_yz()` is pretty much the same thing as their `xy` and `xz` counterparts. I wonder if it would be better to try and re-use much of that code, so that it's easier to maintain.

On the other hand, re-using code like this can make way less readable, so I'm torn. I just wanted to comment this so that we think at least think about it.

@ali-ramadhan what are your thoughts?

Reply:
Also it seems that specifically `test_netcdf_rectilinear_column()` is the only topology where we don't test all three directions...

Reply:
Yeah I also struggled with repeated code vs. less understandable generic code here :( Although I did feel like making it more generic made it quite a bit less readable and debuggable.

Since it was tests where one case can break (e.g. `xy` but not `yz`) I decided to err on the side of repeating code but making it easier to debug and modify specific tests.